### PR TITLE
fix(download): expand ~ to user home dir

### LINKED ISSUE #549: Bug: `deadline job download-output` does not expand ~ when prompting for new root path
### Describe the bug

When downloading job output on a system that does not have the root path format used in the job output files, the CLI prompts the user to enter a new root path. I wanted to save it to `~/tmp` but it ended up saving it to `$PWD/~/tmp`.

```
This root path format does not match the operating system you're using. Where would you like to save the files?
The location was C:\Users\Administrator\test-resources\maya\ClientReleaseTesting, on Windows.
> Please enter a new root path: ~/tmp        

Downloading Outputs  [####################################]  100%          
Download Summary:
    Downloaded 48 files totaling 63.37 MB.
    Total download time of 5.63655 seconds at 11.24 MB/s.
    Download locations (total file counts):
        /Users/jericht/~/~/tmp (48 files)
```

### Expected Behaviour

I expected the `~` to get expanded so that it would save the files to `/Users/jericht/tmp`, not `/Users/jericht/~/tmp`

### Current Behaviour

It looks like it treated the `~/tmp` as a relative path and prepended the current working directory

### Reproduction Steps

Run `deadline job download-output` on a job where it causes this prompt (e.g. a job submitted on Windows and downloaded to Mac)

### Environment

At minimum:
1. Operating system: Windows Server 2022 and Mac
2. Output of `python3 --version`: Python 3.9.6
3. Output of `deadline --version`: deadline, version 0.49.3
4. If this is from a version installed by the Deadline Cloud Submitter installer, then what version of the submitter installer? No

Please share other details about your environment that you think might be relevant to reproducing the bug.

### PR DESCRIPTION
**Pull request by jericht**
Fixes: https://github.com/aws-deadline/deadline-cloud/issues/549

### What was the problem/requirement? (What/Why)

When running `deadline job download-output` on a machine that has a different root path format than the machine a job was submitted from, the CLI prompts the user to enter a new root path to download output files to. If you use a `~`, it does not properly expand that to the user home directory.

### What was the solution? (How)

Expand `~`

### What is the impact of this change?

CLI prompt now expands `~` to user home directory

### How was this change tested?

See [DEVELOPMENT.md](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#testing) for information on running tests.

- Have you run the unit tests?
    - PASSED
- Have you run the integration tests?
    - PASSED
- Have you made changes to the `download` or `asset_sync` modules? If so, then it is highly recommended
  that you ensure that the docker-based unit tests pass.
    - PASSED

Manual testing on Mac. Verified it expands the `~` correctly:

```
Downloading output from Job 'AnimatedSphere.ma (some)'
This root path format does not match the operating system you're using. Where would you like to save the files?
The location was C:\Users\Administrator\test-resources\maya\ClientReleaseTesting, on Windows.
> Please enter a new root path: ~/tmp/files/new

Downloading Outputs  [####################################]  100%          
Download Summary:
    Downloaded 24 files totaling 31.01 MB.
    Total download time of 4.42551 seconds at 7.01 MB/s.
    Download locations (total file counts):
        /Users/jericht/tmp/files/new (24 files)
```

### Was this change documented?

No

### Does this PR introduce new dependencies?

This library is designed to be integrated into third-party applications that have bespoke and customized deployment environments. Adding dependencies will increase the chance of library version conflicts and incompatabilities. Please evaluate the addition of new dependencies. See the [Dependencies](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#dependencies) section of DEVELOPMENT.md for more details.

*   [ ] This PR adds one or more new dependency Python packages. I acknowledge I have reviewed the considerations for adding dependencies in [DEVELOPMENT.md](https://github.com/aws-deadline/deadline-cloud/blob/mainline/DEVELOPMENT.md#dependencies).
*   [x] This PR does not add any new dependencies.

### Is this a breaking change?

No

### Does this change impact security?

No

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/deadline/job_attachments/download.py [hunk 1]
```diff
@@ -1120,7 +1120,7 @@ def set_root_path(self, original_root: str, new_root: str) -> None:
         (It will store the new root path as an absolute path.)
         """
         # Need to use absolute to not resolve symlinks, but need normpath to get rid of relative paths, i.e. '..'
-        new_root = str(os.path.normpath(Path(new_root).absolute()))
+        new_root = str(os.path.normpath(Path(new_root).expanduser().absolute()))
```
Comment (mwiebe) [inline src/deadline/job_attachments/download.py]: This place will call `expanduser()` on all calls to `set_root_path`, but I doubt that's what we want to do. The original issue is about interactive user input, I think it's better to put this call immediately after that input is collected instead of inside this setter.

**Review by mwiebe — CHANGES_REQUESTED** [review]

**Review by sakshie95 — DISMISSED** [review]

File: src/deadline/job_attachments/download.py (hunk 1 above)
Comment (jericht) [reply src/deadline/job_attachments/download.py]: I think we do want it in here since we make a call to `pathlib.Path.absolute()` which will prepend the current working directory to the path in cases where the path starts with `~`, resulting in a root path like `/home/jericht/~/tmp/123`, which is something we don't want in my opinion.

We are applying `os.normpath` and `pathlib.Path.absolute` to the path, so we're already not setting the `new_root` parameter as is to the `self.outputs_by_root` dict. To me, this means the code is intended to "make sure the path is a valid root path". However, transforming a path like `~/tmp/123` into something like `/the/current/working/dir/~/tmp/123`, for example, seems like a bug. If the caller really wants to set a root path like `/the/current/working/dir/~/tmp/123` by just passing in `~/tmp/123` while the cwd is `/the/current/working/dir/`, I think it's reasonable to have the caller pass in the absolute path in this case.

I've updated the function docstring to make it clear that this happens along with the other transformations it applies to `new_root`.

I tested this locally to make sure `expanduser` handles some edge cases I could think of correctly. If the path is already has the `~` expanded, it won't attempt to expand it again creating an invalid path. Here are a few tests I ran locally.

```
>>> def test_expanduser(path: str):
...     print(f"os.path.expanduser: {os.path.expanduser(path)}")
...     print(f"pathlib.Path.expanduser: {pathlib.Path(path).expanduser()}")
... 

>>> test_expanduser("~/tmp/123")
os.path.expanduser: /home/jericht/tmp/123
pathlib.Path.expanduser: /home/jericht/tmp/123

>>> test_expanduser("~/~/tmp/123")
os.path.expanduser: /home/jericht/~/tmp/123
pathlib.Path.expanduser: /home/jericht/~/tmp/123

>>> test_expanduser("/tmp/~/123")
os.path.expanduser: /tmp/~/123
pathlib.Path.expanduser: /tmp/~/123

>>> test_expanduser("/home/jericht/~/tmp/123")
os.path.expanduser: /home/jericht/~/tmp/123
pathlib.Path.expanduser: /home/jericht/~/tmp/123

>>> test_expanduser("~/home/jericht/~/tmp/123")
os.path.expanduser: /home/jericht/home/jericht/~/tmp/123
pathlib.Path.expanduser: /home/jericht/home/jericht/~/tmp/123
```

**Review by leongdl — COMMENTED** [review]
Does this change also affect `deadline attachment download` ? I hope the same logic (or the same code path) will apply

File: src/deadline/job_attachments/download.py (hunk 1 above)
Comment (leongdl) [reply src/deadline/job_attachments/download.py]: side comment: Does this change also work for `deadline attachment download` ?

Comment (mwiebe) [reply src/deadline/job_attachments/download.py]: If we do want this expansion to happen always in the setter, I think we should look more closely at where the value comes from currently, and where it could come from. Will this be surprising in any case?

**Comment by godobyte** [comment]
> Does this change also affect `deadline attachment download` ? I hope the same logic (or the same code path) will apply

`deadline attachment download` does not use this `set_root_path` API. It downloads to specified destination in path mapping file, or the working directory if it's not provided. Maybe it's worth checking the path mapping approach by defining `~` as part of destination in the input file. Upon checking with Jericho, this path normalization issue is related to click, whereas `attachment download` doesn't use user prompt path for download.

File: src/deadline/job_attachments/download.py (hunk 1 above)
Comment (jericht) [reply src/deadline/job_attachments/download.py]: After thinking about it, I want to take a safer approach and only fix what the issue called out originally. I've pushed a change to only expand the `~` in that specific case.

**Review by mwiebe — DISMISSED** [review]

**Review by mwiebe — APPROVED** [review]

**Review by leongdl — APPROVED** [review]
